pkg/controllers/disruption/orchestration/queue.go

```diff
@@ -0,0 +1,386 @@
+		// 3. Remove it from the Queue's internal data structure
+		failedLaunches := lo.Filter(cmd.ReplacementKeys, func(key *NodeClaimKey, _ int) bool {
+			return !key.Initialized
+		})
+		disruptionReplacementNodeClaimFailedCounter.With(map[string]string{
+			methodLabel:            cmd.Method,
+			consolidationTypeLabel: cmd.ConsolidationType,
+		}).Add(float64(len(failedLaunches)))
+		q.cluster.UnmarkForDeletion(lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string { return s.ProviderID() })...)
+		err = multierr.Append(err, q.RequireNoScheduleTaint(ctx, false, cmd.Candidates...))
+		err = multierr.Append(err, cmd.LastError)
+		nodeNames := strings.Join(lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string {
+			return s.Name()
+		}), ",")
+		q.Remove(cmd)
+		return reconcile.Result{RequeueAfter: controller.Immediately}, fmt.Errorf("failed to disrupt nodes %s, %w", nodeNames, err)
+	}
+	// Requeue command if not complete
+	return reconcile.Result{RequeueAfter: controller.Immediately}, nil
+}
+
+func (q *Queue) Process(ctx context.Context, cmd *Command) error {
+	if q.clock.Since(cmd.TimeAdded) > maxRetryDuration {
+		return NewUnrecoverableError(fmt.Errorf("command at %s reached timeout after %s", cmd.TimeAdded, q.clock.Since(cmd.TimeAdded)))
+	}
+	// If the time hasn't expired, either wait or terminate.
+	if err := q.WaitOrTerminate(ctx, cmd); err != nil {
+		// If there was an error, set this as the command's last error so that we can propagate it.
+		if IsUnrecoverableError(err) {
+			return fmt.Errorf("got unrecoverable error, %w", err)
+		}
+		return err
+	}
+	return nil
+}
+
+// WaitOrTerminate will wait until launched nodeclaims are ready.
+// Once the replacements are ready, it will terminate the candidates.
+// Will return true if the item in the queue should be re-queued. If a command has
+// timed out, this will return false.
+// nolint:gocyclo
+func (q *Queue) WaitOrTerminate(ctx context.Context, cmd *Command) error {
+	waitErrs := make([]error, len(cmd.ReplacementKeys))
+	for i := range cmd.ReplacementKeys {
+		key := cmd.ReplacementKeys[i]
+		// If we know the node claim is initialized, no need to check again.
+		if key.Initialized {
+			continue
+		}
+		// Get the nodeclaim
+		nodeClaim, err := nodeclaim.Get(ctx, q.kubeClient, key.Key)
+		if err != nil {
+			// The NodeClaim got deleted after an initial eventual consistency delay
+			// This means that there was an ICE error or the Node initializationTTL expired
+			// In this case, the error is unrecoverable, so don't requeue.
+			if apierrors.IsNotFound(err) {
+				return NewUnrecoverableError(err)
+			}
+			waitErrs[i] = fmt.Errorf("getting node claim, %w", err)
+			continue
+		}
+		// We emitted this event when disruption was blocked on launching/termination.
+		// This does not block other forms of deprovisioning, but we should still emit this.
+		q.recorder.Publish(disruptionevents.Launching(nodeClaim, cmd.Reason))
```

Comment:
Should we emit this again here? Or is this a one-time event that we do?

Reply:
We only do one launch, but we have a de-dupe period here. I could go either way, since the event de-dupe will catch this just fine.